plugins/permission-backend/src/service/router.ts

```diff
@@ -0,0 +1,148 @@
+  AuthorizeResult,
+  AuthorizeResponse,
+  AuthorizeRequest,
+  Identified,
+} from '@backstage/plugin-permission-common';
+import { PermissionPolicy } from '@backstage/plugin-permission-node';
+import { PermissionIntegrationClient } from './PermissionIntegrationClient';
+
+/**
+ * Options required when constructing a new {@link express#Router} using
+ * {@link createRouter}.
+ *
+ * @public
+ */
+export interface RouterOptions {
+  logger: Logger;
+  discovery: PluginEndpointDiscovery;
+  policy: PermissionPolicy;
+}
+
+const handleRequest = async (
+  { id, resourceRef, ...request }: Identified<AuthorizeRequest>,
+  user: BackstageIdentity | undefined,
+  policy: PermissionPolicy,
+  permissionIntegrationClient: PermissionIntegrationClient,
+  authHeader?: string,
+): Promise<Identified<AuthorizeResponse>> => {
+  const response = await policy.handle(request, user);
+
+  if (response.result === AuthorizeResult.CONDITIONAL) {
+    // Sanity check that any resource provided matches the one expected by the permission
+    if (request.permission.resourceType !== response.conditions.resourceType) {
+      throw new Error(
+        `Invalid resource conditions returned from permission policy for permission ${request.permission.name}`,
+      );
+    }
+
+    if (resourceRef) {
+      return {
+        id,
+        ...(await permissionIntegrationClient.applyConditions(
+          {
+            resourceRef,
+            ...response.conditions,
+          },
+          authHeader,
+        )),
+      };
+    }
+
+    return {
+      id,
+      result: AuthorizeResult.CONDITIONAL,
+      // TODO(mtlewis): this .conditions.conditions situation is a bit awkward. I think it's
+      // worth exploring a bit of reorganization of the ConditionalPolicyResult type so that
+      // the naming of property chains like this makes a bit more sense.
+      conditions: response.conditions.conditions,
+    };
+  }
+
+  return { id, ...response };
+};
+
+/**
+ * Creates a new {@link express#Router} which provides the backend API
+ * for the permission system.
+ *
+ * @public
+ */
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { policy, discovery } = options;
+
+  const identity = new IdentityClient({
+    discovery,
+    issuer: await discovery.getExternalBaseUrl('auth'),
+  });
+
+  const permissionIntegrationClient = new PermissionIntegrationClient({
+    discovery,
+  });
+
+  const router = Router();
+  router.use(express.json());
+
+  router.get('/health', (_, response) => {
+    response.send({ status: 'ok' });
+  });
+
+  router.post(
+    '/authorize',
+    async (
+      req: Request<Identified<AuthorizeRequest>[]>,
+      res: Response<Identified<AuthorizeResponse>[]>,
+    ) => {
+      const token = IdentityClient.getBearerToken(req.header('authorization'));
+      const user = token ? await identity.authenticate(token) : undefined;
+
+      const body: Identified<AuthorizeRequest>[] = req.body;
```

Comment:
no zod parsing of this one?

Reply:
Good spot, thanks! Added in fafb482d73. Also noticed when working on this that the router test suite was a bit sparse, added more tests in 2bacc8be34.